storage/src/mmr/mem.rs

```diff
@@ -0,0 +1,352 @@
+//! A bare-bones MMR structure without pruning and where all nodes are hashes & maintained in
+//! memory.
+
+use crate::mmr::{Hash, Hasher, Proof};
+
+/// Implementation of `InMemoryMMR`.
+pub struct InMemoryMMR<const N: usize, H: Hasher<N>> {
+    hasher: H,
+    // The nodes of the MMR, laid out according to a post-order traversal of the MMR trees, starting
+    // from the from highest peak to lowest.
+    nodes: Vec<Hash<N>>,
+}
+
+/// A PeakIterator returns a (height, position) tuple for each peak in the MMR, in decreasing order
+/// of height. Height starts at 0 for leaves, and a peak position is its 0-based offset into the
+/// underlying nodes vector. The iterator will only return the peaks that existed at the time of its
+/// initialization if new elements are added to the MMR between iterations.
+#[derive(Default)]
+struct PeakIterator {
+    sz: u64,    // Number of nodes in the MMR at the point the iterator was initialized.
+    peak: u64,  // 1-based index of the current peak, which may not be in the MMR.
+    two_h: u64, // 2^(height+1) of the current peak, where a leaf has a height of 0.
+}
+
+impl Iterator for PeakIterator {
+    type Item = (u32, u64); // (height, peak)
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.two_h > 1 {
+            if self.peak <= self.sz {
+                // found a peak
+                let r = (self.two_h.trailing_zeros() - 1, self.peak - 1);
+                // move to the right sibling
+                self.peak += self.two_h - 1;
+                assert!(self.peak > self.sz); // sibling shouldn't be in the MMR if MMR is valid
+                return Some(r);
+            }
+            // descend to the left child
+            self.two_h >>= 1;
+            self.peak -= self.two_h;
+        }
+        None
+    }
+}
+
+// Return a new iterator over the peaks of a MMR with the given number of nodes.
+fn peak_iterator(sz: u64) -> PeakIterator {
+    if sz == 0 {
+        return PeakIterator::default();
```

Comment:
Not super familiar with the construction but should `two_h = 1` (instead of `0`) here? Just based on the comment:
```
// 2^(height+1) of the current peak, where a leaf has a height of 0.
```

Where I assume 2-3 nodes has height of 1... 1 node has a height of 0... and 0 nodes has a height of 0 or -1?

Reply:
I guess I think of this as degenerate case .. an empty MMR basically has undefined height, since height=0 is used for leaves. Either 0 or 1 has the same effect where the iterator will produce no values.